File: app/src/main/java/de/wollis_page/gibsonos/activity/LoginActivity.java
package de.wollis_page.gibsonos.activity;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;

import de.wollis_page.gibsonos.R;
import de.wollis_page.gibsonos.activity.base.GibsonOsActivity;
import de.wollis_page.gibsonos.model.Account;
import de.wollis_page.gibsonos.task.UserTask;

public class LoginActivity extends GibsonOsActivity {

    private List<EditText> editTexts;
    private EditText etAlias, etUser, etPassword, etUrl;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        etAlias = findViewById(R.id.alias);
        etUser = findViewById(R.id.user);
        etPassword = findViewById(R.id.password);
        etUrl = findViewById(R.id.url);

        editTexts = new ArrayList<>();
        editTexts.add(etAlias);
        editTexts.add(etUser);
        editTexts.add(etPassword);
        editTexts.add(etUrl);

        etAlias.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                List<Account> accounts = Account.find(Account.class, "alias = ?", s.toString());

                if (accounts.size() > 0) {
                    etAlias.setError(getString(R.string.account_error_alias_exists));
                }
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        final LoginActivity activity = this;
        Button btnLogin = findViewById(R.id.login);
        btnLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
            boolean hasErrors = false;

            for (EditText editText : editTexts) {
                if (editText.getText().toString().length() == 0) {
                    editText.setError(getString(R.string.account_error_value_empty));
                }

                if (editText.getError() != null && editText.getError().toString().length() > 0) {
                    hasErrors = true;
                }
            }

            if (hasErrors) {
                return;
            }

            CompletableFuture.supplyAsync(new Supplier<Object>() {
                @Override
                public Object get() {
                    //v.setEnabled(false);
                    Account account = UserTask.login(
                        activity,
                        etUrl.getText().toString(),
                        etUser.getText().toString(),
                        etPassword.getText().toString()
                    );
                    assert account != null;
                    account.setAlias(etAlias.getText().toString());
                    account.setUrl(etUrl.getText().toString());
                    account.save();

                    //v.setEnabled(true);
                    setResult(RESULT_OK);
                    activity.finish();

                    return null;
                }
            });
            }
        });
    }
}

File: app/src/main/java/de/wollis_page/gibsonos/activity/LoginActivity.kt
package de.wollis_page.gibsonos.activity

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import com.orm.SugarRecord
import de.wollis_page.gibsonos.R
import de.wollis_page.gibsonos.activity.base.GibsonOsActivity
import de.wollis_page.gibsonos.exception.MessageException
import de.wollis_page.gibsonos.helper.Config
import de.wollis_page.gibsonos.model.Account
import de.wollis_page.gibsonos.task.UserTask
import java.util.*
import java.util.concurrent.CompletableFuture

class LoginActivity : GibsonOsActivity() {
    private var editTexts: MutableList<EditText?> = ArrayList()
    private var etAlias: EditText? = null
    private var etUser: EditText? = null
    private var etPassword: EditText? = null
    private var etUrl: EditText? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_login)
        super.onCreate(savedInstanceState)
        val me = this

        me.etAlias = me.findViewById(R.id.alias) as EditText
        me.etUser = me.findViewById(R.id.user) as EditText
        me.etPassword = me.findViewById(R.id.password) as EditText
        me.etUrl = me.findViewById(R.id.url) as EditText

        me.editTexts = ArrayList()
        me.editTexts.add(me.etAlias)
        me.editTexts.add(me.etUser)
        me.editTexts.add(me.etPassword)
        me.editTexts.add(etUrl)

        me.etAlias?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val accounts: List<Account> = SugarRecord.find(Account::class.java, "alias = ?", s.toString())

                if (accounts.isNotEmpty()) {
                    me.etAlias?.error = getString(R.string.account_error_alias_exists)
                }
            }

            override fun afterTextChanged(s: Editable) {}
        })

        val btnLogin = findViewById<Button>(R.id.login)

        btnLogin.setOnClickListener OnClickListener@{
            var hasErrors = false

            for (editText in me.editTexts as ArrayList<EditText?>) {
                if (editText?.text.isNullOrBlank()) {
                    editText?.error = getString(R.string.account_error_value_empty)
                }

                if (!editText?.error.isNullOrEmpty()) {
                    hasErrors = true
                }
            }

            if (hasErrors) {
                Log.d(Config.LOG_TAG, "Login has errors")

                return@OnClickListener
            }

            CompletableFuture.supplyAsync<Any> {
                try {
                    val account = UserTask.login(
                        me,
                        me.etUrl?.text.toString(),
                        me.etUser?.text.toString(),
                        me.etPassword?.text.toString()
                    )
                    account.alias = me.etAlias?.text.toString()
                    account.save()
                    me.application.addAccount(account)

                    setResult(RESULT_OK)
                    me.finish()
                } catch (exception: MessageException) {
                    me.runOnUiThread {
                        var message = exception.message
                        val messageRessource = exception.messageRessource

                        if (messageRessource != null) {
                            message = getString(messageRessource)
                        }

                        Toast.makeText(me, message, Toast.LENGTH_SHORT).show()
                    }
                }
            }.exceptionally { e -> e.printStackTrace() }
        }
    }
}